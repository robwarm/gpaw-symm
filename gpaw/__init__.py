# Copyright (C) 2003  CAMP
# Please see the accompanying LICENSE file for further information.

"""Main gpaw module."""

import os
import sys
try:
    from distutils.util import get_platform
except ImportError:
    modulepath = os.environ.get('GPAW_GET_PLATFORM')
    if modulepath is None:
        errmsg = ('Error: Could not get platform from distutils. '
                  'Set the GPAW_GET_PLATFORM environment variable to '
                  'the architecture string printed during build.')
        raise ImportError(errmsg)

    def get_platform():
        return modulepath

from glob import glob
from os.path import join, isfile

import numpy as np

assert not np.version.version.startswith('1.6.0')


__all__ = ['GPAW', 'Calculator',
           'Mixer', 'MixerSum', 'MixerDif', 'MixerSum2',
           'CG', 'Davidson', 'RMM_DIIS', 'LCAO',
           'PoissonSolver',
           'FermiDirac', 'MethfesselPaxton',
           'restart']


class ConvergenceError(Exception):
    pass


class KohnShamConvergenceError(ConvergenceError):
    pass


class PoissonConvergenceError(ConvergenceError):
    pass


# Check for special command line arguments:
debug = False
trace = False
dry_run = 0
memory_estimate_depth = 2
parsize_domain = None
parsize_bands = None
sl_default = None
sl_diagonalize = None
sl_inverse_cholesky = None
sl_lcao = None
sl_lrtddft = None
buffer_size = None
extra_parameters = {}
profile = False
i = 1
while len(sys.argv) > i:
    arg = sys.argv[i]
    if arg.startswith('--gpaw-'):
        # Found old-style gpaw command line argument:
        arg = '--' + arg[7:]
        raise RuntimeError('Warning: Use %s instead of %s.' %
                           (arg, sys.argv[i]))
    if arg == '--trace':
        trace = True
    elif arg == '--debug':
        debug = True
    elif arg.startswith('--dry-run'):
        dry_run = 1
        if len(arg.split('=')) == 2:
            dry_run = int(arg.split('=')[1])
    elif arg.startswith('--memory-estimate-depth'):
        memory_estimate_depth = -1
        if len(arg.split('=')) == 2:
            memory_estimate_depth = int(arg.split('=')[1])
    elif arg.startswith('--domain-decomposition='):
        parsize_domain = [int(n) for n in arg.split('=')[1].split(',')]
        if len(parsize_domain) == 1:
            parsize_domain = parsize_domain[0]
        else:
            assert len(parsize_domain) == 3
    elif arg.startswith('--state-parallelization='):
        parsize_bands = int(arg.split('=')[1])
    elif arg.startswith('--sl_default='):
        # --sl_default=nprow,npcol,mb,cpus_per_node
        # use 'd' for the default of one or more of the parameters
        # --sl_default=default to use all default values
        sl_args = [n for n in arg.split('=')[1].split(',')]
        if len(sl_args) == 1:
            assert sl_args[0] == 'default'
            sl_default = ['d'] * 3
        else:
            sl_default = []
            assert len(sl_args) == 3
            for sl_args_index in range(len(sl_args)):
                assert sl_args[sl_args_index] is not None
                if sl_args[sl_args_index] is not 'd':
                    assert int(sl_args[sl_args_index]) > 0
                    sl_default.append(int(sl_args[sl_args_index]))
                else:
                    sl_default.append(sl_args[sl_args_index])
    elif arg.startswith('--sl_diagonalize='):
        # --sl_diagonalize=nprow,npcol,mb,cpus_per_node
        # use 'd' for the default of one or more of the parameters
        # --sl_diagonalize=default to use all default values
        sl_args = [n for n in arg.split('=')[1].split(',')]
        if len(sl_args) == 1:
            assert sl_args[0] == 'default'
            sl_diagonalize = ['d'] * 3
        else:
            sl_diagonalize = []
            assert len(sl_args) == 3
            for sl_args_index in range(len(sl_args)):
                assert sl_args[sl_args_index] is not None
                if sl_args[sl_args_index] is not 'd':
                    assert int(sl_args[sl_args_index]) > 0
                    sl_diagonalize.append(int(sl_args[sl_args_index]))
                else:
                    sl_diagonalize.append(sl_args[sl_args_index])
    elif arg.startswith('--sl_inverse_cholesky='):
        # --sl_inverse_cholesky=nprow,npcol,mb,cpus_per_node
        # use 'd' for the default of one or more of the parameters
        # --sl_inverse_cholesky=default to use all default values
        sl_args = [n for n in arg.split('=')[1].split(',')]
        if len(sl_args) == 1:
            assert sl_args[0] == 'default'
            sl_inverse_cholesky = ['d'] * 3
        else:
            sl_inverse_cholesky = []
            assert len(sl_args) == 3
            for sl_args_index in range(len(sl_args)):
                assert sl_args[sl_args_index] is not None
                if sl_args[sl_args_index] is not 'd':
                    assert int(sl_args[sl_args_index]) > 0
                    sl_inverse_cholesky.append(int(sl_args[sl_args_index]))
                else:
                    sl_inverse_cholesky.append(sl_args[sl_args_index])
    elif arg.startswith('--sl_lcao='):
        # --sl_lcao=nprow,npcol,mb,cpus_per_node
        # use 'd' for the default of one or more of the parameters
        # --sl_lcao=default to use all default values
        sl_args = [n for n in arg.split('=')[1].split(',')]
        if len(sl_args) == 1:
            assert sl_args[0] == 'default'
            sl_lcao = ['d'] * 3
        else:
            sl_lcao = []
            assert len(sl_args) == 3
            for sl_args_index in range(len(sl_args)):
                assert sl_args[sl_args_index] is not None
                if sl_args[sl_args_index] is not 'd':
                    assert int(sl_args[sl_args_index]) > 0
                    sl_lcao.append(int(sl_args[sl_args_index]))
                else:
                    sl_lcao.append(sl_args[sl_args_index])
    elif arg.startswith('--sl_lrtddft='):
        # --sl_lcao=nprow,npcol,mb,cpus_per_node
        # use 'd' for the default of one or more of the parameters
        # --sl_lcao=default to use all default values
        sl_args = [n for n in arg.split('=')[1].split(',')]
        if len(sl_args) == 1:
            assert sl_args[0] == 'default'
            sl_lrtddft = ['d'] * 3
        else:
            sl_lrtddft = []
            assert len(sl_args) == 3
            for sl_args_index in range(len(sl_args)):
                assert sl_args[sl_args_index] is not None
                if sl_args[sl_args_index] is not 'd':
                    assert int(sl_args[sl_args_index]) > 0
                    sl_lrtddft.append(int(sl_args[sl_args_index]))
                else:
                    sl_lrtddft.append(sl_args[sl_args_index])
    elif arg.startswith('--buffer_size='):
        # Buffer size for MatrixOperator in MB
        buffer_size = int(arg.split('=')[1])
    elif arg.startswith('--gpaw='):
        extra_parameters = eval('dict(%s)' % arg[7:])
    elif arg == '--gpaw':
        extra_parameters = eval('dict(%s)' % sys.argv.pop(i + 1))
    elif arg.startswith('--profile='):
        profile = arg.split('=')[1]
    else:
        i += 1
        continue
    # Delete used command line argument:
    del sys.argv[i]

if debug:
    np.seterr(over='raise', divide='raise', invalid='raise', under='ignore')

    oldempty = np.empty

    def empty(*args, **kwargs):
        a = oldempty(*args, **kwargs)
        try:
            a.fill(np.nan)
        except ValueError:
            a.fill(-1000000)
        return a
    np.empty = empty


build_path = join(__path__[0], '..', 'build')
arch = '%s-%s' % (get_platform(), sys.version[0:3])

# If we are running the code from the source directory, then we will
# want to use the extension from the distutils build directory:
sys.path.insert(0, join(build_path, 'lib.' + arch))


def get_gpaw_python_path():
    paths = os.environ['PATH'].split(os.pathsep)
    paths.insert(0, join(build_path, 'bin.' + arch))
    for path in paths:
        if isfile(join(path, 'gpaw-python')):
            return path
    raise RuntimeError('Could not find gpaw-python!')


try:
    setup_paths = os.environ['GPAW_SETUP_PATH'].split(os.pathsep)
except KeyError:
    if os.pathsep == ';':
        setup_paths = [r'C:\gpaw-setups']
    else:
        setup_paths = ['/usr/local/share/gpaw-setups',
                       '/usr/share/gpaw-setups']


from gpaw.aseinterface import GPAW
from gpaw.mixer import Mixer, MixerSum, MixerDif, MixerSum2
from gpaw.eigensolvers import Davidson, RMM_DIIS, CG, LCAO
from gpaw.poisson import PoissonSolver
from gpaw.occupations import FermiDirac, MethfesselPaxton
from gpaw.wavefunctions.pw import PW


class Calculator(GPAW):
    def __init__(self, *args, **kwargs):
        sys.stderr.write('Please start using GPAW instead of Calculator!\n')
        GPAW.__init__(self, *args, **kwargs)


def restart(filename, Class=GPAW, **kwargs):
    calc = Class(filename, **kwargs)
    atoms = calc.get_atoms()
    return atoms, calc


if trace:
    indent = '    '
    path = __path__[0]
    from gpaw.mpi import parallel, rank
    if parallel:
        indent = 'CPU%d    ' % rank

    def f(frame, event, arg):
        global indent
        f = frame.f_code.co_filename
        if not f.startswith(path):
            return

        if event == 'call':
            print('%s%s:%d(%s)' % (indent, f[len(path):], frame.f_lineno,
                                   frame.f_code.co_name))
            indent += '| '
        elif event == 'return':
            indent = indent[:-2]

    sys.setprofile(f)


if profile:
    from cProfile import Profile
    import atexit
    prof = Profile()

    def f(prof, filename):
        prof.disable()
        from gpaw.mpi import rank
        if filename == '-':
            prof.print_stats('time')
        else:
            prof.dump_stats(filename + '.%04d' % rank)
    atexit.register(f, prof, profile)
    prof.enable()


command = os.environ.get('GPAWSTARTUP')
if command is not None:
    exec(command)

    
def is_parallel_environment():
    """Check if we are running in a parallel environment.
    
    This function can be redefined in ~/.gpaw/rc.py.  Example::
        
        def is_parallel_environment():
            import os
            return 'PBS_NODEFILE' in os.environ
    """
    return False
    
    
home = os.environ.get('HOME')
if home is not None:
    rc = os.path.join(home, '.gpaw', 'rc.py')
    if os.path.isfile(rc):
        # Read file in ~/.gpaw/rc.py
        execfile(rc)
