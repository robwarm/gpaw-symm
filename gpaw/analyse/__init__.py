# python wants to have this file
