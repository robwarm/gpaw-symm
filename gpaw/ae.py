"""Bare Coulomb potential for hydrogen."""

import numpy as np

from gpaw.utilities import erf
from gpaw.spline import Spline
from gpaw.setup import BaseSetup
from gpaw.basis_data import Basis


class HydrogenAllElectronSetup(BaseSetup):

    def __init__(self, alpha1=10.0, alpha2=300.0):
        self.alpha1 = alpha1
        self.alpha2 = alpha2
        
        self.natoms = 0
        self.E = 0.0
        self.Z = 1
        self.Nc = 0
        self.Nv = 1
        self.nao = None
        self.pt_j = []
        self.ni = 0
        self.l_j = []
        self.n_j = []
        self.nct = Spline(0, 0.5, [0.0, 0.0, 0.0])
        self.Nct = 0.0
        rc = 2.0
        r_g = np.linspace(0, rc, 100)
        r2_g = r_g**2
        self.ghat_l = [Spline(0, rc, 4 * alpha1**1.5 / np.pi**0.5 *
                              np.exp(-alpha1 * r2_g))]
        v_g = erf(alpha1**0.5 * r_g) - erf(alpha2**0.5 * r_g)
        v_g[1:] *= (4 * np.pi)**0.5 / r_g[1:]
        v_g[0] = 4 * (alpha1**0.5 - alpha2**0.5)
        self.vbar = Spline(0, rc, v_g)
        self.Delta_pL = np.zeros((0, 1))
        self.Delta0 = -1 / (4 * np.pi)**0.5
        self.lmax = 0
        self.K_p = self.M_p = self.MB_p = np.zeros(0)
        self.M_pp = np.zeros((0, 0))
        self.Kc = 0.0
        self.MB = 0.0
        self.M = -(alpha1 / 2 / np.pi)**0.5
        self.xc_correction = None
        self.HubU = None
        self.dO_ii = np.zeros((0, 0))
        self.type = 'all-electron'
        self.fingerprint = None
        
    def get_default_nbands(self):
        return 1
        
    def build(self, basis):
        if basis is None:
            basis = Basis('H', 'sz(dzp)')
        elif isinstance(basis, str):
            basis = Basis('H', basis)
        self.basis = basis
        self.phit_j = self.basis.tosplines()
        self.f_j = [1.0]
        self.nao = self.basis.nao
    
    def print_info(self, text):
        text('Hydrogen all-electron potential')
