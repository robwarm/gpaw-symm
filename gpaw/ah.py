"""Appelbaum-Hamann local pseudo-potential for silicon.

See::

  Self-Consistent Pseudopotential for Si
  Joel A. Appelbaum and D. R. Hamann
  PRB 8, 1777 (1973)
  
"""

import numpy as np

from gpaw.setup import BaseSetup
from gpaw.spline import Spline
from gpaw.basis_data import Basis


class AppelbaumHamann(BaseSetup):
    def __init__(self, alpha=0.6102, v1=3.042, v2=-1.372):
        self.alpha = alpha
        self.v1 = v1
        self.v2 = v2
        
        self.E = 0.0
        self.Z = 14
        self.Nc = 10
        self.Nv = 4
        self.nao = None
        nullspline = Spline(0, 0.5, [0., 0., 0.])
        self.pt_j = [nullspline]
        self.ni = 1
        self.l_j = [0]
        self.f_j = [4]
        self.n_j = [1]
        self.nct = nullspline
        self.Nct = 0.0
        rc = 4.0
        r2_g = np.linspace(0, rc, 100)**2
        x_g = np.exp(-alpha * r2_g)
        self.ghat_l = [Spline(0, rc, 4 * alpha**1.5 / np.pi**0.5 * x_g)]
        self.vbar = Spline(0, rc, 2 * np.pi**0.5 * (v1 + v2 * r2_g) * x_g)
        self.Delta_pL = np.zeros((1, 1))
        self.Delta0 = -4 / (4 * np.pi)**0.5
        self.lmax = 0
        self.K_p = self.M_p = self.MB_p = np.zeros(1)
        self.M_pp = np.zeros((1, 1))
        self.Kc = 0.0
        self.MB = 0.0
        self.M = 0.0
        self.xc_correction = None
        self.HubU = None
        self.dO_ii = np.zeros((1, 1))
        self.type = 'ah'
        self.fingerprint = None
        
    def build(self, basis):
        if basis is None:
            basis = Basis('Si', 'sz(dzp)')
        elif isinstance(basis, str):
            basis = Basis('Si', basis)
        self.basis = basis
        self.phit_j = self.basis.tosplines()
        self.nao = self.basis.nao
    
    def print_info(self, text):
        text('Appelbaum-Hamann pseudo potential')
        
    def calculate_initial_occupation_numbers(self, magmom, hund, charge,
                                             nspins):
        assert nspins == 1
        return np.array([(2.0, 2.0 / 3, 2.0 / 3, 2.0 / 3)])

    def initialize_density_matrix(self, f_si):
        return np.zeros((len(f_si), 1))

    def get_default_nbands(self):
        return 3
