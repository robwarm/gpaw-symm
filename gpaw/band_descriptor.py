# Copyright (C) 2003  CAMP
# Please see the accompanying LICENSE file for further information.

"""Band-descriptors for blocked/strided groups.

This module contains classes defining two kinds of band groups:

* Blocked groups with contiguous band indices.
* Strided groups with evenly-spaced band indices.
"""

import numpy as np

from gpaw.mpi import serial_comm

NONBLOCKING = False

class BandDescriptor:
    """Descriptor-class for ordered lists of bands

    A ``BandDescriptor`` object holds information on how functions, such
    as wave functions and corresponding occupation numbers, are divided
    into groups according to band indices. The main information here is
    how many bands are stored on each processor and who gets what.

    This is how a 12 band array is laid out in memory on 3 cpu's::

      a) Blocked groups       b) Strided groups

           3    7   11             9   10   11
     myn   2 \  6 \ 10       myn   6    7    8
      |    1  \ 5  \ 9        |    3    4    5
      |    0    4    8        |    0    1    2
      |                       |
      +----- band_rank        +----- band_rank

    Example:

     >>> a = np.zeros((3, 4))
     >>> a.ravel()[:] = range(12)
     >>> a
     array([[ 0,  1,  2,  3],
            [ 4,  5,  6,  7],
            [ 8,  9, 10, 11]])
     >>> b = np.zeros((4, 3))
     >>> b.ravel()[:] = range(12)
     >>> b.T
     array([[ 0,  3,  6,  9],
            [ 1,  4,  7, 10],
            [ 2,  5,  8, 11]])
     """
    
    def __init__(self, nbands, comm=None, strided=False):
        """Construct band-descriptor object.

        Parameters:

        nbands: int
            Global number of bands.
        comm: MPI-communicator
            Communicator for band-groups.
        strided: bool
            Enable strided band distribution for better
            load balancing with many unoccupied bands.

        Note that if comm.size is 1, then all bands are contained on
        a single CPU and blocked/strided grouping loses its meaning.

        Attributes:

        ============  ======================================================
        ``nbands``    Number of bands in total.
        ``mynbands``  Number of bands on this CPU.
        ``beg``       Beginning of band indices in group (inclusive).
        ``end``       End of band indices in group (exclusive).
        ``step``      Stride for band indices between ``beg`` and ``end``.
        ``comm``      MPI-communicator for band distribution.
        ============  ======================================================
        """
        
        if comm is None:
            comm = serial_comm
        self.comm = comm
        self.rank = self.comm.rank

        self.nbands = nbands

        if self.nbands % self.comm.size != 0:
            raise RuntimeError('Cannot distribute %d bands to %d processors' %
                               (self.nbands, self.comm.size))

        self.mynbands = self.nbands // self.comm.size
        self.strided = strided

        nslice = self.get_slice()
        self.beg, self.end, self.step = nslice.indices(self.nbands)

    def __len__(self):
        return self.mynbands

    def get_slice(self, band_rank=None):
        """Return the slice of global bands which belong to a given rank."""
        if band_rank is None:
            band_rank = self.comm.rank
        assert band_rank in xrange(self.comm.size)

        if self.strided:
            nstride = self.comm.size
            nslice = slice(band_rank, None, nstride)
        else:
            n0 = band_rank * self.mynbands
            nslice = slice(n0, n0 + self.mynbands)
        return nslice

    def get_band_indices(self, band_rank=None):
        """Return the global band indices which belong to a given rank."""
        nslice = self.get_slice(band_rank)
        return np.arange(*nslice.indices(self.nbands))

    def get_band_ranks(self):
        """Return array of ranks as a function of global band indices."""
        rank_n = np.empty(self.nbands, dtype=int)
        for band_rank in range(self.comm.size):
            nslice = self.get_slice(band_rank)
            rank_n[nslice] = band_rank
        assert (rank_n >= 0).all() and (rank_n < self.comm.size).all()
        return rank_n

    def who_has(self, n):
        """Convert global band index to rank information and local index."""
        if self.strided:
            myn, band_rank = divmod(n, self.comm.size)
        else:
            band_rank, myn = divmod(n, self.mynbands)
        return band_rank, myn

    def global_index(self, myn, band_rank=None):
        """Convert rank information and local index to global index."""
        if band_rank is None:
            band_rank = self.comm.rank
        if self.strided:
            n = band_rank + myn * self.comm.size
        else:
            n = band_rank * self.mynbands + myn
        return n

    def get_size_of_global_array(self):
        return (self.nbands,)

    def zeros(self, n=(), dtype=float, global_array=False):
        """Return new zeroed 3D array for this domain.

        The type can be set with the ``dtype`` keyword (default:
        ``float``).  Extra dimensions can be added with ``n=dim``.
        A global array spanning all domains can be allocated with
        ``global_array=True``."""
        #TODO XXX doc
        return self._new_array(n, dtype, True, global_array)
    
    def empty(self, n=(), dtype=float, global_array=False):
        """Return new uninitialized 3D array for this domain.

        The type can be set with the ``dtype`` keyword (default:
        ``float``).  Extra dimensions can be added with ``n=dim``.
        A global array spanning all domains can be allocated with
        ``global_array=True``."""
        #TODO XXX doc
        return self._new_array(n, dtype, False, global_array)
        
    def _new_array(self, n=(), dtype=float, zero=True, global_array=False):
        if global_array:
            shape = self.get_size_of_global_array()
        else:
            shape = (self.mynbands,)
            
        if isinstance(n, int):
            n = (n,)

        shape = tuple(shape) + n

        if zero:
            return np.zeros(shape, dtype)
        else:
            return np.empty(shape, dtype)

    def collect(self, a_nx, broadcast=False):
        """Collect distributed array to master-CPU or all CPU's."""
        if self.comm.size == 1:
            return a_nx

        xshape = a_nx.shape[1:]

        # Optimization for blocked groups
        if not self.strided:
            if broadcast:
                A_nx = self.empty(xshape, a_nx.dtype, global_array=True)
                self.comm.all_gather(a_nx, A_nx)
                return A_nx

            if self.rank == 0:
                A_nx = self.empty(xshape, a_nx.dtype, global_array=True)
            else:
                A_nx = None
            self.comm.gather(a_nx, 0, A_nx)
            return A_nx

        # Collect all arrays on the master:
        if self.rank != 0:
            # There can be several sends before the corresponding receives
            # are posted, so use syncronous send here
            self.comm.ssend(a_nx, 0, 3011)
            if broadcast:
                A_nx = self.empty(xshape, a_nx.dtype, global_array=True)
                self.comm.broadcast(A_nx, 0)
                return A_nx
            else:
                return None

        # Put the band groups from the slaves into the big array
        # for the whole collection of bands:
        A_nx = self.empty(xshape, a_nx.dtype, global_array=True)
        for band_rank in range(self.comm.size):
            if band_rank != 0:
                a_nx = self.empty(xshape, a_nx.dtype, global_array=False)
                self.comm.receive(a_nx, band_rank, 3011)
            A_nx[self.get_slice(band_rank), ...] = a_nx

        if broadcast:
            self.comm.broadcast(A_nx, 0)
        return A_nx

    def distribute(self, B_nx, b_nx):
        """ distribute full array B_nx to band groups, result in
        b_nx. b_nx must be allocated."""

        if self.comm.size == 1:
            b_nx[:] = B_nx
            return

        # Optimization for blocked groups
        if not self.strided:
            self.comm.scatter(B_nx, b_nx, 0)
            return

        if self.rank != 0:
            self.comm.receive(b_nx, 0, 421)
            return
        else:
            requests = []
            for band_rank in range(self.comm.size):
                if band_rank != 0:
                    a_nx = B_nx[self.get_slice(band_rank), ...].copy()
                    request = self.comm.send(a_nx, band_rank, 421, NONBLOCKING)
                    # Remember to store a reference to the
                    # send buffer (a_nx) so that is isn't
                    # deallocated:
                    requests.append((request, a_nx))
                else:
                    b_nx[:] = B_nx[self.get_slice(), ...]
                        
            for request, a_nx in requests:
                self.comm.wait(request)

