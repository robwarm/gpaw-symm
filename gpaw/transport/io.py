#This module is used to store and read some temperary data.
import cPickle
import os
from gpaw.mpi import world
from gpaw.transport.tools import collect_atomic_matrices, gather_ndarray_dict
import numpy as np

class Transport_IO:
    def __init__(self, kpt_comm, domain_comm):
        self.kpt_comm = kpt_comm
	self.domain_comm = domain_comm
	assert self.kpt_comm.size * self.domain_comm.size == world.size
	self.dir_name = 'temperary_data'
        if world.rank == 0:
            if not os.access(self.dir_name, os.F_OK):
                os.mkdir(self.dir_name)
        world.barrier()
        self.filenames = self.default_temperary_filenames()

    def default_temperary_filenames(self):
        filenames = {}
	for name in ['Lead']:
	    filenames[name] = name
	for i in range(self.kpt_comm.size):
	    for j in range(self.domain_comm.size):
	        name = 'KC_' + str(i) + '_DC_' + str(j) +'AD'
		#Local analysis data on kpt_comm i and domain_comm j
		filenames[name] = name
	return filenames

    def read_data(self, bias_step=0, filename=None, option='Analysis'):
        if option == 'Lead':
	    if filename is None:
	        filename = self.filenames[option]
  	    fd = file(self.dir_name + '/' + filename, 'r')
	    data = cPickle.load(fd)
	    fd.close()
	elif option == 'Analysis':
            name = 'KC_' + str(self.kpt_comm.rank) + '_DC_' + \
	                                     str(self.domain_comm.rank) +'AD'
	    fd = file(self.dir_name + '/' + self.filenames[name] + '_bias_' + str(bias_step), 'r')
	    data = cPickle.load(fd)
	    fd.close()
	else:
	    raise NotImplementError
    	return data
           
    def save_data(self, obj, bias_step=0, filename=None, option='Analysis'):
        #       option ------map------ obj
        #       Analysis            Transport.Analysor
	#        Lead                Lead_Calc
        data = self.collect_data(obj, option)
	if option == 'Lead':
	    if world.rank == 0:
	        if filename is None:
		    filename = self.filenames[option]
  	        fd = file(self.dir_name + '/' + filename, 'wb')
		cPickle.dump(data, fd, 2)
		fd.close()
	elif option == 'Analysis':
	    name = 'KC_' + str(self.kpt_comm.rank) + '_DC_' + \
	                                     str(self.domain_comm.rank) +'AD'
	    fd = file(self.dir_name + '/' + self.filenames[name] + '_bias_' + str(bias_step), 'wb')
	    cPickle.dump(data, fd, 2)
	    fd.close()
	else:
	    raise NotImplementError()
	
    def collect_data(self, obj, option):
        if option == 'Lead':
	    data = self.collect_lead_data(obj)
	elif option == 'Analysis':
	    data = self.collect_analysis_data(obj)
	else:
	    raise NotImplementError
	return data

    def collect_lead_data(self, obj): 	
        data = {}
        data['bzk_kc'] = obj.wfs.bzk_kc
        #data['bzk_qc'] = obj.wfs.bzk_qc
        data['ibzk_kc'] = obj.wfs.ibzk_kc
        data['ibzk_qc'] = obj.wfs.ibzk_qc
        #data['setups'] = obj.wfs.setups
	data['cell_cv'] = obj.gd.cell_cv
	data['parsize_c'] = obj.gd.parsize_c
	data['pbc_c'] = obj.gd.pbc_c
        data['N_c'] = obj.gd.N_c
	data['nao'] = obj.wfs.setups.nao
        data['fine_N_c'] = obj.finegd.N_c
        data['nspins'] = obj.wfs.nspins
	data['fermi'] = obj.get_fermi_level()

        den, ham = obj.density, obj.hamiltonian
        gd = obj.gd
        vt_sG = gd.collect(ham.vt_sG)
        nt_sG = gd.collect(den.nt_sG)
        data['vt_sG'] = vt_sG
        data['nt_sG'] = nt_sG

        finegd = obj.finegd
        vt_sg = finegd.collect(ham.vt_sg)
        nt_sg = finegd.collect(den.nt_sg)
        vHt_g = finegd.collect(ham.vHt_g)
        rhot_g = finegd.collect(den.rhot_g)
        data['vt_sg'] = vt_sg
        data['nt_sg'] = nt_sg
        data['vHt_g'] = vHt_g
        data['rhot_g'] = rhot_g

        D_asp = collect_atomic_matrices(den.D_asp, den.setups,
                                        den.nspins, den.gd.comm,
            			    den.rank_a)
        dH_asp = collect_atomic_matrices(ham.dH_asp, ham.setups,
                                         ham.nspins, ham.gd.comm,
            			     ham.rank_a)
        data['D_asp'] = D_asp
        data['dH_asp'] = dH_asp
        return data

    def collect_analysis_data(self, obj):
        return obj.data
        
    def arrange_analysis_data(self, n_bias_step, n_ion_step, analysis_mode):           
        data = self.read_data(bias_step=n_bias_step, option='Analysis')
	parsize = data['domain_parsize']
        parpos = data['domain_parpos']
	domain_rank = data['domain_rank']
	kpt_rank = data['kpt_rank']
	kpt_size = data['kpt_size']
	transmission_dimension = data['transmission_dimension']
	dos_dimension = data['dos_dimension']
        assert kpt_rank == self.kpt_comm.rank
	assert domain_rank == self.domain_comm.rank

        #collect transmission, dos, current
	global_data = {}
	flag = 'K_' + str(kpt_rank) + 'D_' + str(domain_rank) + '_'
        global_data[flag + 'parpos'] = data['domain_parpos']
        global_data[flag + 'tc'] = data['tc']
	global_data[flag + 'dos'] = data['dos']

        global_data = gather_ndarray_dict(global_data, self.kpt_comm)
	
	global_data[flag + 'vt'] = data['vt']
	global_data[flag + 'vtx'] = data['vtx']
	global_data[flag + 'vty'] = data['vty']
	global_data[flag + 'nt'] = data['nt']
	global_data[flag + 'ntx'] = data['ntx']
	global_data[flag + 'nty'] = data['nty']

        global_data = gather_ndarray_dict(global_data, self.domain_comm)

        global_data['lead_fermi'] = data['lead_fermi']
	global_data['bias'] = data['bias']
	global_data['gate'] = data['gate']
	global_data['charge'] = data['charge']
	global_data['magmom'] = data['magmom']
	global_data['local_magmom'] = data['local_magmom']

	global_data['newform'] = True
        global_data['domain_parsize'] = data['domain_parsize']
	global_data['kpt_size'] = data['kpt_size']
	global_data['transmission_dimension'] = data['transmission_dimension']
	global_data['dos_dimension'] = data['dos_dimension']

	world.barrier()
        if world.rank == 0:
            if analysis_mode:
                filename = '/abias_step_' + str(n_bias_step)
            else:
                filename = '/bias_step_' + str(n_bias_step)
            fd = file('analysis_data/ionic_step_' + str(n_ion_step)
                      + filename, 'wb')
            cPickle.dump(global_data, fd, 2)
            fd.close()
	    for root, dirs, files in os.walk('temperary_data'):
	        for name in files:
		    if 'AD' in name:
		        os.remove(os.path.join(root, name))
	        
       

	
     
            


	
        

	       
	    
        

