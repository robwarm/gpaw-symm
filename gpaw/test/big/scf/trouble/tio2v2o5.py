from ase.io import read
from gpaw import GPAW
atoms = read('tio2v2o5.txt')
atoms.calc = GPAW(h=0.20,
                  kpts = (2,2,1),
                  xc='RPBE')
ncpus = 8
