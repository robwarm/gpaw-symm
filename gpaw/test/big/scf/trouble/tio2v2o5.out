
  ___ ___ ___ _ _ _  
 |   |   |_  | | | | 
 | | | | | . | | | | 
 |__ |  _|___|_____|  0.9.1.11315
 |___|_|             


Positions:
  0 O     8.3411    1.9309    9.2647
  1 O     3.2338    0.0854    9.4461
  2 Ti    1.5783    0.1417    8.4327
  3 Ti    6.6126    2.0588    8.2126
  4 Ti    9.4072    2.0891    7.6857
  5 Ti    4.2748    0.1256    7.9470
  6 O     9.9477    0.2283    7.3281
  7 O     4.7391    2.0618    7.6111
  8 O     7.7533    2.1354    6.7529
  9 O     2.6347    0.2585    6.9403
 10 O     6.2080    0.1462    8.6280
 11 O     0.9517    1.9798    8.9832
 12 O     8.5835    6.2280    9.4350
 13 O     3.1930    3.9040    9.7886
 14 Ti    1.4899    4.0967    8.8898
 15 Ti    6.6167    5.8865    8.8463
 16 Ti    9.3207    5.9258    7.7482
 17 Ti    4.1984    3.9386    8.2442
 18 O     9.9075    3.9778    7.6337
 19 O     4.7626    5.8322    8.0051
 20 O     7.6143    5.6963    7.1276
 21 O     2.5760    3.9661    7.3115
 22 O     6.2303    3.8223    8.8067
 23 O     1.1298    5.9913    8.6968
 24 O     8.3845    9.7338    9.1214
 25 O     3.1730    7.9593    9.3632
 26 Ti    1.5914    7.8120    8.2310
 27 Ti    6.7003    9.7064    8.1528
 28 Ti    9.3943    9.7202    7.6037
 29 Ti    4.3168    7.7857    7.9666
 30 O     9.9045    7.7968    7.2716
 31 O     4.7772    9.7015    7.4648
 32 O     7.7314    9.7221    6.6253
 33 O     2.7673    7.6929    6.8222
 34 O     6.2358    7.8628    8.6557
 35 O     1.0528    9.7017    8.5919
 36 V     8.4820    5.0952   11.4981
 37 V     9.7787    2.0447   11.0800
 38 O     6.4427    5.6315   10.7415
 39 O    10.7389    0.4065   11.8697
 40 O     8.3109    3.0048   12.4083
 41 O    10.4702    4.1612   10.6543
 42 O     8.9827    6.3884   13.0109

        .------------------------.  
       /|                        |  
      / |                        |  
     /  |                        |  
    /   |                        |  
   /    |                        |  
  /     |                        |  
 *      |                        |  
 |      |                        |  
 |      |                  O     |  
 |      |                        |  
 |      |                        |  
 |      |              O V  O    |  
 |      | O   O     O  Ti        |  
 | OO   | Ti       O O    V   Ti |  
 |     O|  O    Ti   Ti   O   O  |  
 |     Ti    O  O  O        Ti   |  
 |   O  |     Ti       O    O    |  
 |      |OO        Ti    Ti      |  
 |   Ti |   TiO O                |  
 |      O            O    O      |  
 |      |                        |  
 |      .------------------------.  
 |     /                        /   
 |    /                        /    
 |   /                        /     
 |  /                        /      
 | /                        /       
 |/                        /        
 *------------------------*         

Unit Cell:
           Periodic     X           Y           Z      Points  Spacing
  --------------------------------------------------------------------
  1. axis:    yes   10.152054    0.000000    0.000000    52     0.1952
  2. axis:    yes    0.000000   11.430000    0.000000    56     0.2041
  3. axis:    no     0.000000    0.000000   18.295483    92     0.1989

